File: java_solutions/318.maximum-product-of-word-lengths.java
/*
 * @lc app=leetcode id=318 lang=java
 *
 * [318] Maximum Product of Word Lengths
 */

class Solution {
    public int maxProduct(String[] words) {
        return Optional.of(Arrays.stream(words).map(w -> new int[]{w.chars().reduce(0, (m, c) -> m | (1 << (c - 'a'))), w.length()}).toArray(int[][]::new)).map(data -> IntStream.range(0, data.length).flatMap(i -> IntStream.range(i + 1, data.length).filter(j -> (data[i][0] & data[j][0]) == 0).map(j -> data[i][1] * data[j][1])).max().orElse(0)).get();
    }
}

File: kotlin_solutions/318.maximum-product-of-word-lengths.kt
/*
 * @lc app=leetcode id=318 lang=kotlin
 *
 * [318] Maximum Product of Word Lengths
 */

class Solution {fun maxProduct(words: Array<String>): Int = words.indices.associate { it to words[it].toSet() }.let { map -> words.indices.flatMap { i -> ((i + 1) until words.size).map { j -> i to j } }.filter { (i, j) -> words[i].none { c -> map[j]!!.contains(c) } }.maxOfOrNull { (i, j) -> words[i].length * words[j].length } ?: 0 }}
